components/scream/src/physics/spa/spa_functions_impl.hpp

```diff
@@ -520,27 +527,29 @@ ::update_spa_data_from_file(
-  typename view_1d<Real>::HostMirror PS_v_h("PS",spa_horiz_interp.source_grid_ncols);
-  typename view_2d<Real>::HostMirror CCN3_v_h("CCN3",spa_horiz_interp.source_grid_ncols,source_data_nlevs);
-  typename view_3d<Real>::HostMirror AER_G_SW_v_h("AER_G_SW",spa_horiz_interp.source_grid_ncols,nswbands,source_data_nlevs);
-  typename view_3d<Real>::HostMirror AER_SSA_SW_v_h("AER_SSA_SW",spa_horiz_interp.source_grid_ncols,nswbands,source_data_nlevs);
-  typename view_3d<Real>::HostMirror AER_TAU_SW_v_h("AER_TAU_SW",spa_horiz_interp.source_grid_ncols,nswbands,source_data_nlevs);
-  typename view_3d<Real>::HostMirror AER_TAU_LW_v_h("AER_TAU_LW",spa_horiz_interp.source_grid_ncols,nlwbands,source_data_nlevs);
+  typename view_1d<Real>::HostMirror PS_v_h("PS",num_local_cols);
+  typename view_2d<Real>::HostMirror CCN3_v_h("CCN3",num_local_cols,source_data_nlevs);
+  typename view_3d<Real>::HostMirror AER_G_SW_v_h("AER_G_SW",num_local_cols,nswbands,source_data_nlevs);
+  typename view_3d<Real>::HostMirror AER_SSA_SW_v_h("AER_SSA_SW",num_local_cols,nswbands,source_data_nlevs);
+  typename view_3d<Real>::HostMirror AER_TAU_SW_v_h("AER_TAU_SW",num_local_cols,nswbands,source_data_nlevs);
+  typename view_3d<Real>::HostMirror AER_TAU_LW_v_h("AER_TAU_LW",num_local_cols,nlwbands,source_data_nlevs);
 
   // Construct the grid needed for input:
-  auto grid = std::make_shared<PointGrid>("grid",spa_horiz_interp.source_grid_ncols,source_data_nlevs,self_comm);
-  PointGrid::dofs_list_type dof_gids("",spa_horiz_interp.source_grid_ncols);
-  Kokkos::parallel_for("", spa_horiz_interp.source_grid_ncols, KOKKOS_LAMBDA (const int& ii) {
-    dof_gids(ii) = ii;
-  });
+  auto grid = std::make_shared<PointGrid>("grid",num_local_cols,source_data_nlevs,comm);
+  PointGrid::dofs_list_type dof_gids("",num_local_cols);
+  auto dof_gids_h = Kokkos::create_mirror_view(dof_gids);
+  for (const auto& nn : spa_horiz_interp.source_local_col_map) {
+    dof_gids_h(nn.second) = nn.first;
+  }
+  Kokkos::deep_copy(dof_gids,dof_gids_h);
   grid->set_dofs(dof_gids);
 
   using namespace ShortFieldTagsNames;
   FieldLayout scalar1d_layout { {LEV}, {source_data_nlevs} };
-  FieldLayout scalar2d_layout_mid { {COL}, {spa_horiz_interp.source_grid_ncols} };
-  FieldLayout scalar3d_layout_mid { {COL,LEV}, {spa_horiz_interp.source_grid_ncols, source_data_nlevs} };
-  FieldLayout scalar3d_swband_layout { {COL,SWBND, LEV}, {spa_horiz_interp.source_grid_ncols, nswbands, source_data_nlevs} }; 
-  FieldLayout scalar3d_lwband_layout { {COL,LWBND, LEV}, {spa_horiz_interp.source_grid_ncols, nlwbands, source_data_nlevs} };
+  FieldLayout scalar2d_layout_mid { {COL}, {num_local_cols} };
+  FieldLayout scalar3d_layout_mid { {COL,LEV}, {num_local_cols, source_data_nlevs} };
+  FieldLayout scalar3d_swband_layout { {COL,SWBND, LEV}, {num_local_cols, nswbands, source_data_nlevs} }; 
+  FieldLayout scalar3d_lwband_layout { {COL,LWBND, LEV}, {num_local_cols, nlwbands, source_data_nlevs} };
```

Comment:
I put the comment here, cause github doesn't let me comment below.

I think this function is way more expensive that it should be. Besides the grid construction, we have lots of views allocations, as well as the AtmosphereInput construction/setup (with all the scorpio stuff). I think we should be able to avoid these.

Yes, this routine is only called once per month (actually twice), but still. Why making the call more expensive? Do we know how much does a SPA call cost when the month has changed, compared to when we're in the middle of it? That might tell us if we're dramatically inefficient (and need to fix this) or if the overhead is negligible in the grand scheme of things.